### ISSUE TITLE
zippath.child('..') does not raise InsecurePath

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4535]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4535|
|Type|defect|
|Created|2010-06-30 14:52:27Z|
|Branch|https://github.com/twisted/twisted/tree/zippath-child-insecure-4535|

```
>>> from twisted.python.zippath import ZipArchive
>>> a = ZipArchive('subtext.zip')
>>> a.child('subtext').child('..')
ZipPath('/home/exarkun/subtext.zip')
>>> a.child('subtext').child('..').path
'subtext.zip/subtext/..'
>>> 
```

This isn't quite as catastrophic as it first appears, since:

```
>>> a.child('subtext').child('..').open()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/exarkun/Projects/Twisted/branches/child-args-3169-2/twisted/python/zippath.py", line 113, in open
    return self.archive.zipfile.readfile(self.pathInArchive)
  File "/home/exarkun/Projects/Twisted/branches/child-args-3169-2/twisted/python/zipstream.py", line 33, in readfile
    zinfo = self.getinfo(name)
  File "/usr/lib/python2.6/zipfile.py", line 821, in getinfo
    'There is no item named %r in the archive' % name)
KeyError: "There is no item named 'subtext/..' in the archive"
>>> 
```

However code using the `path` attribute of one of these directly may still be vulnerable to some sort of attack.

And this behavior clearly violates the intent of the `child` method.

<details><summary>Searchable metadata</summary>

```
trac-id__4535 4535
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_zippath_child_insecure_4535 branches/zippath-child-insecure-4535
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1277909547000000 1277909547000000
changetime__1279028467000000 1279028467000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [29506]) Branching to 'zippath-child-insecure-4535'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Okay, so, rejecting ".." and "CON" and suchlike would restrict access to legitimate zipfile contents, so this behavior isn't exactly incorrect.

In the branch, I've added a note to `ZipPath.child` about this difference in behavior.  I've also correctly the grossly incorrect handling of ".." in `ZipPath.__repr__` which was mostly responsible for the confusing results above - ".." in a zip archive path doesn't mean "parent directory".  It means ".." and that's it.

The tests (and, hey, implementation) end up rather gross because ZipPath tries to use platform-native path conventions.  If someone can do better, please feel free.

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/zippath-child-insecure-4535)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @exarkun|
|-|-|

[1]
+        path = os.sep.join(parts)

Don't you want to use os.path.join here? I guess it makes no difference?

[2]
+        string literals are esacped in the ZipPath repr

Typo: escaped.

+1!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Thanks!

> Don't you want to use os.path.join here? I guess it makes no difference?

I guess `os.sep.join` is better because you can have a file in a zip archive named "//////////" if you want.  `os.path.join` would get confused.

> Typo: escaped.

Fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [29797]) Merge zippath-child-insecure-4535

Author: exarkun
Reviewer: therve
Fixes: [#4535](https://github.com/twisted/twisted/issues/4535)

Change the repr string of `twisted.python.zippath.ZipPath` instances to more
faithfully represent the underlying path.  Previously, incorrect platform-specific
processing would be applied to the path string before returning it from the
repr method, causing it to be misleading in the presence of ".." (and possible
other) path components.

Also document the potentially surprising behavior of `ZipPath.child` with
respect to special (or rather, not special) path components such as "..".

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
